File: pkg/trajectory_ir/runtime/log.py
import contextlib
import json
import sqlite3
import threading
from typing import Any

from trajectory_ir.runtime.nodes import Node


class SlotConflictError(ValueError):
    """Raised when a different payload is written to an occupied (trajectory, step,
    seq, kind) slot."""


class NodeLog:
    """Append-only, content-addressed node log backed by SQLite.

    Appends are idempotent: replaying an append for a node whose id already
    exists is a no-op (`INSERT OR IGNORE`). This is what makes DBOS's
    workflow replay safe to layer on top of -- re-running an already-appended
    step produces the same node id and is silently absorbed instead of
    duplicating history, so there is no separate "seal" operation needed.

    Concurrent writers are serialized with a lock and ``BEGIN IMMEDIATE`` so
    the block-and-gate claim path cannot double-run a non-idempotent tool.
    """

    def __init__(self, db_path: str):
        # check_same_thread=False: the durable backend replays a crashed
        # workflow on its own recovery thread, so the log must be writable from a
        # thread other than the one that opened it. Access is serialized via
        # self._lock; only one thread mutates the connection at a time.
        self._lock = threading.RLock()
        self._conn = sqlite3.connect(db_path, check_same_thread=False)
        self._conn.execute(
            """
            CREATE TABLE IF NOT EXISTS nodes (
                id TEXT PRIMARY KEY,
                trajectory_id TEXT NOT NULL,
                tenant_id TEXT NOT NULL,
                step_n INTEGER,
                seq INTEGER NOT NULL,
                kind TEXT NOT NULL,
                payload_json TEXT NOT NULL,
                ts REAL NOT NULL
            )
            """,
        )
        # Logical slot uniqueness per tenant: same content id is still
        # INSERT OR IGNORE; a different payload at the same slot is rejected.
        self._conn.execute(
            """
            CREATE UNIQUE INDEX IF NOT EXISTS idx_nodes_slot
            ON nodes (tenant_id, trajectory_id, IFNULL(step_n, -1), seq, kind)
            """,
        )
        self._conn.execute(
            """
            CREATE INDEX IF NOT EXISTS idx_nodes_traj_tenant
            ON nodes (trajectory_id, tenant_id)
            """,
        )
        self._conn.commit()

    def append(
        self,
        kind: str,
        step_n: int | None,
        payload: dict,
        trajectory_id: str,
        tenant_id: str,
        seq: int,
    ) -> Node:
        node = Node(
            kind=kind,
            trajectory_id=trajectory_id,
            tenant_id=tenant_id,
            step_n=step_n,
            seq=seq,
            payload=payload,
        )
        with self._lock:
            self._conn.execute(
                "INSERT OR IGNORE INTO nodes "
                "(id, trajectory_id, tenant_id, step_n, seq, kind, payload_json, ts) "
                "VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
                (
                    node.id,
                    node.trajectory_id,
                    node.tenant_id,
                    node.step_n,
                    node.seq,
                    node.kind,
                    json.dumps(node.payload),
                    node.ts,
                ),
            )
            # INSERT OR IGNORE also swallows unique-slot conflicts. Detect a
            # different payload occupying this logical slot and fail closed.
            cur = self._conn.execute(
                "SELECT id FROM nodes WHERE tenant_id = ? AND trajectory_id = ? "
                "AND IFNULL(step_n, -1) = IFNULL(?, -1) AND seq = ? AND kind = ?",
                (tenant_id, trajectory_id, step_n, seq, kind),
            )
            row = cur.fetchone()
            if row is not None and row[0] != node.id:
                self._conn.rollback()
                raise SlotConflictError(
                    f"slot conflict trajectory={trajectory_id!r} step={step_n} "
                    f"seq={seq} kind={kind!r}: different payload already stored"
                )
            self._conn.commit()
        return node

    def claim_tool_call(
        self,
        step_n: int,
        payload: dict,
        trajectory_id: str,
        tenant_id: str,
        seq: int,
    ) -> bool:
        """Atomically claim a TOOL_CALL slot.

        Returns True if this caller inserted the TOOL_CALL (may run the tool).
        Returns False if a TOOL_CALL already occupied this (trajectory, step, seq).
        """
        node = Node(
            kind="TOOL_CALL",
            trajectory_id=trajectory_id,
            tenant_id=tenant_id,
            step_n=step_n,
            seq=seq,
            payload=payload,
        )
        with self._lock:
            try:
                self._conn.execute("BEGIN IMMEDIATE")
                cur = self._conn.execute(
                    "SELECT 1 FROM nodes WHERE tenant_id = ? AND trajectory_id = ? "
                    "AND step_n = ? AND kind = ? AND seq = ? LIMIT 1",
                    (tenant_id, trajectory_id, step_n, "TOOL_CALL", seq),
                )
                if cur.fetchone() is not None:
                    self._conn.execute("COMMIT")
                    return False
                self._conn.execute(
                    "INSERT INTO nodes "
                    "(id, trajectory_id, tenant_id, step_n, seq, kind, payload_json, ts) "
                    "VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
                    (
                        node.id,
                        node.trajectory_id,
                        node.tenant_id,
                        node.step_n,
                        node.seq,
                        node.kind,
                        json.dumps(node.payload),
                        node.ts,
                    ),
                )
                self._conn.execute("COMMIT")
                return True
            except sqlite3.IntegrityError:
                self._conn.execute("ROLLBACK")
                return False
            except Exception:
                with contextlib.suppress(sqlite3.Error):
                    self._conn.execute("ROLLBACK")
                raise

    def has(
        self,
        trajectory_id: str,
        tenant_id: str,
        step_n: int,
        kind: str,
        seq: int | None = None,
    ) -> bool:
        """Does a node of `kind` exist for this trajectory/step scoped to `tenant_id`?

        `tenant_id` is required to enforce multi-tenant isolation. Callers that
        genuinely need cross-tenant existence checks (e.g., administrative or
        diagnostic tools) must use `has_all_tenants` explicitly instead.

        `seq` narrows the question to one exact node slot. Without it, a step
        containing several tool calls cannot distinguish them: one completed
        call's TOOL_RESULT would answer for a different, interrupted call. Left
        as None the query is unscoped, which is the right question for
        step-level kinds (DECISION, COMMIT_STEP) that occur at most once.
        """
        if not tenant_id:
            raise ValueError("tenant_id must be a non-empty string")
        return self._has(trajectory_id, tenant_id=tenant_id, step_n=step_n, kind=kind, seq=seq)

    def has_all_tenants(
        self,
        trajectory_id: str,
        step_n: int,
        kind: str,
        seq: int | None = None,
    ) -> bool:
        """Does a node of `kind` exist for this trajectory/step across any tenant?

        This bypasses tenant isolation and should only be used for administrative
        diagnostics or debugging tools, not standard execution paths.
        """
        return self._has(trajectory_id, tenant_id=None, step_n=step_n, kind=kind, seq=seq)

    def _has(
        self,
        trajectory_id: str,
        *,
        tenant_id: str | None,
        step_n: int,
        kind: str,
        seq: int | None,
    ) -> bool:
        sql = "SELECT 1 FROM nodes WHERE trajectory_id = ? AND step_n = ? AND kind = ?"
        params: list[object] = [trajectory_id, step_n, kind]
        if tenant_id is not None:
            sql += " AND tenant_id = ?"
            params.append(tenant_id)
        if seq is not None:
            sql += " AND seq = ?"
            params.append(seq)
        with self._lock:
            cur = self._conn.execute(sql + " LIMIT 1", params)
            return cur.fetchone() is not None

    def count(self, node_id: str) -> int:
        with self._lock:
            cur = self._conn.execute("SELECT COUNT(*) FROM nodes WHERE id = ?", (node_id,))
            return int(cur.fetchone()[0])

    def list_nodes(
        self,
        trajectory_id: str,
        *,
        tenant_id: str,
    ) -> list[dict[str, Any]]:
        """Return stored nodes for a trajectory ordered by step then seq.

        ``tenant_id`` is required so this method always enforces tenant
        isolation at the read path. Callers that genuinely need rows across
        tenants (e.g. ``export_tir``'s own mixed-tenant check) must use
        ``list_nodes_all_tenants`` explicitly instead.
        """
        return self._list_nodes(trajectory_id, tenant_id=tenant_id)

    def list_nodes_all_tenants(self, trajectory_id: str) -> list[dict[str, Any]]:
        """Return stored nodes for a trajectory across all tenants.

        This bypasses tenant isolation and should only be used by code that
        performs its own tenant-scoping check on the result, such as
        ``export_tir`` detecting a trajectory with mixed ``tenant_id`` values.
        """
        return self._list_nodes(trajectory_id, tenant_id=None)

    def _list_nodes(
        self,
        trajectory_id: str,
        *,
        tenant_id: str | None,
    ) -> list[dict[str, Any]]:
        sql = """
            SELECT id, trajectory_id, tenant_id, step_n, seq, kind, payload_json, ts
            FROM nodes
            WHERE trajectory_id = ?
            """
        params: list[object] = [trajectory_id]
        if tenant_id is not None:
            sql += " AND tenant_id = ?"
            params.append(tenant_id)
        sql += " ORDER BY COALESCE(step_n, -1), seq"
        with self._lock:
            cur = self._conn.execute(sql, params)
            rows: list[dict[str, Any]] = []
            for row in cur.fetchall():
                rows.append(
                    {
                        "id": row[0],
                        "trajectory_id": row[1],
                        "tenant_id": row[2],
                        "step_n": row[3],
                        "seq": row[4],
                        "kind": row[5],
                        "payload": json.loads(row[6]),
                        "ts": row[7],
                    }
                )
            return rows

    def close(self):
        with self._lock:
            self._conn.close()

    def __del__(self):
        # Client SDK call sites construct a fresh NodeLog per function call
        # and never close it explicitly. Without this, connections only get
        # released whenever the garbage collector happens to get to them,
        # which over a long running agent session means an unbounded number
        # of open SQLite connections. This is a best-effort backstop, not a
        # substitute for callers using close()/a context manager when they
        # can.
        with contextlib.suppress(Exception):
            self._conn.close()

File: go/trajir/log/log.go
// Package nodelog provides an append only, content addressed node log on SQLite.
// Behaviour matches pkg/trajectory_ir/runtime/log.py so Go and Python share one IR story.
//
// The import path is trajir/log; the package name is nodelog to avoid clashing with the
// standard library log package at call sites.
package nodelog

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"sync"

	"github.com/Coder-s-OG-s/Trajectory-IR/go/trajir/nodes"
	"github.com/Coder-s-OG-s/Trajectory-IR/go/trajir/postgres"

	_ "modernc.org/sqlite"
)

var ErrSlotConflict = postgres.ErrSlotConflict

// NodeLog is an append only SQLite log keyed by content addressed node id.
// INSERT OR IGNORE makes replaying the same node a no op, which is what durable
// step replay relies on.
type NodeLog struct {
	db *sql.DB
	mu sync.Mutex
}

// Open creates or opens a SQLite database at path and ensures the schema exists.
func Open(path string) (*NodeLog, error) {
	db, err := sql.Open("sqlite", path)
	if err != nil {
		return nil, fmt.Errorf("open sqlite: %w", err)
	}
	// One connection is enough for the local profile and keeps write order simple.
	db.SetMaxOpenConns(1)

	if _, err := db.Exec(`PRAGMA journal_mode=WAL;`); err != nil {
		_ = db.Close()
		return nil, fmt.Errorf("set wal: %w", err)
	}

	schema := `
CREATE TABLE IF NOT EXISTS nodes (
    id TEXT PRIMARY KEY,
    trajectory_id TEXT NOT NULL,
    tenant_id TEXT NOT NULL,
    step_n INTEGER,
    seq INTEGER NOT NULL,
    kind TEXT NOT NULL,
    payload_json TEXT NOT NULL,
    ts REAL NOT NULL
);
CREATE UNIQUE INDEX IF NOT EXISTS idx_nodes_slot
ON nodes (tenant_id, trajectory_id, IFNULL(step_n, -1), seq, kind);
CREATE INDEX IF NOT EXISTS idx_nodes_traj_tenant
ON nodes (trajectory_id, tenant_id);
`
	if _, err := db.Exec(schema); err != nil {
		_ = db.Close()
		return nil, fmt.Errorf("create schema: %w", err)
	}

	return &NodeLog{db: db}, nil
}

// Append builds a node via trajir/nodes and stores it. Same content id is
// ignored (idempotent replay). Different payload at the same logical slot
// returns ErrSlotConflict.
func (l *NodeLog) Append(
	kind string,
	stepN *int,
	payload map[string]any,
	trajectoryID, tenantID string,
	seq int,
) (*nodes.Node, error) {
	n, err := nodes.NewNode(kind, trajectoryID, tenantID, stepN, seq, payload)
	if err != nil {
		return nil, err
	}

	payloadJSON, err := json.Marshal(n.Payload)
	if err != nil {
		return nil, fmt.Errorf("marshal payload: %w", err)
	}

	var step any
	if n.StepN != nil {
		step = *n.StepN
	}

	l.mu.Lock()
	defer l.mu.Unlock()

	res, err := l.db.Exec(
		`INSERT OR IGNORE INTO nodes
		 (id, trajectory_id, tenant_id, step_n, seq, kind, payload_json, ts)
		 VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
		n.ID,
		n.TrajectoryID,
		n.TenantID,
		step,
		n.Seq,
		n.Kind,
		string(payloadJSON),
		n.TS,
	)
	if err != nil {
		return nil, fmt.Errorf("insert node: %w", err)
	}

	rowsAffected, err := res.RowsAffected()
	if err != nil {
		return nil, fmt.Errorf("rows affected: %w", err)
	}
	if rowsAffected == 1 {
		return n, nil
	}

	owner, err := l.slotOwner(tenantID, trajectoryID, step, seq, kind)
	if err != nil {
		return nil, fmt.Errorf("slot owner: %w", err)
	}
	if owner != "" && owner != n.ID {
		return nil, fmt.Errorf("%w: trajectory=%s step=%v seq=%d kind=%s",
			ErrSlotConflict, trajectoryID, step, seq, kind)
	}
	return n, nil
}

func (l *NodeLog) slotOwner(tenantID, trajectoryID string, step any, seq int, kind string) (string, error) {
	var id string
	err := l.db.QueryRow(
		`SELECT id FROM nodes
		 WHERE tenant_id = ? AND trajectory_id = ?
		   AND IFNULL(step_n, -1) = IFNULL(?, -1)
		   AND seq = ? AND kind = ?`,
		tenantID, trajectoryID, step, seq, kind,
	).Scan(&id)
	if err == sql.ErrNoRows {
		return "", nil
	}
	return id, err
}

// ClaimToolCall atomically inserts a TOOL_CALL at (trajectory, step, seq).
// Returns claimed=true if this caller owns the slot and may run the tool.
// Returns claimed=false if the slot was already taken (block-and-gate).
func (l *NodeLog) ClaimToolCall(
	stepN int,
	payload map[string]any,
	trajectoryID, tenantID string,
	seq int,
) (claimed bool, err error) {
	n, err := nodes.NewNode("TOOL_CALL", trajectoryID, tenantID, &stepN, seq, payload)
	if err != nil {
		return false, err
	}
	payloadJSON, err := json.Marshal(n.Payload)
	if err != nil {
		return false, fmt.Errorf("marshal payload: %w", err)
	}

	l.mu.Lock()
	defer l.mu.Unlock()

	tx, err := l.db.Begin()
	if err != nil {
		return false, err
	}
	defer func() {
		if err != nil {
			_ = tx.Rollback()
		}
	}()

	var one int
	qerr := tx.QueryRow(
		`SELECT 1 FROM nodes WHERE tenant_id = ? AND trajectory_id = ? AND step_n = ? AND kind = ? AND seq = ? LIMIT 1`,
		tenantID, trajectoryID, stepN, "TOOL_CALL", seq,
	).Scan(&one)
	if qerr == nil {
		if err = tx.Commit(); err != nil {
			return false, err
		}
		return false, nil
	}
	if qerr != sql.ErrNoRows {
		err = qerr
		return false, err
	}

	_, err = tx.Exec(
		`INSERT INTO nodes
		 (id, trajectory_id, tenant_id, step_n, seq, kind, payload_json, ts)
		 VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
		n.ID,
		n.TrajectoryID,
		n.TenantID,
		stepN,
		n.Seq,
		n.Kind,
		string(payloadJSON),
		n.TS,
	)
	if err != nil {
		// Unique slot conflict under race: treat as not claimed.
		_ = tx.Rollback()
		return false, nil
	}
	if err = tx.Commit(); err != nil {
		return false, err
	}
	return true, nil
}

// Has reports whether a node of kind exists for trajectory and step scoped to tenantID.
// When seq is non nil, the match is limited to that sequence slot.
// tenantID must be non-empty; use HasAllTenants for cross-tenant administrative checks.
func (l *NodeLog) Has(trajectoryID, tenantID string, stepN int, kind string, seq *int) (bool, error) {
	if tenantID == "" {
		return false, fmt.Errorf("tenantID must be a non-empty string")
	}
	return l.has(trajectoryID, &tenantID, stepN, kind, seq)
}

// HasAllTenants reports whether a node of kind exists for trajectory and step across any tenant.
// This bypasses tenant isolation and is intended for administrative or diagnostic tools only.
func (l *NodeLog) HasAllTenants(trajectoryID string, stepN int, kind string, seq *int) (bool, error) {
	return l.has(trajectoryID, nil, stepN, kind, seq)
}

func (l *NodeLog) has(trajectoryID string, tenantID *string, stepN int, kind string, seq *int) (bool, error) {
	l.mu.Lock()
	defer l.mu.Unlock()

	q := `SELECT 1 FROM nodes WHERE trajectory_id = ? AND step_n = ? AND kind = ?`
	args := []any{trajectoryID, stepN, kind}
	if tenantID != nil {
		q += ` AND tenant_id = ?`
		args = append(args, *tenantID)
	}
	if seq != nil {
		q += ` AND seq = ?`
		args = append(args, *seq)
	}
	q += ` LIMIT 1`

	var one int
	err := l.db.QueryRow(q, args...).Scan(&one)
	if err == sql.ErrNoRows {
		return false, nil
	}
	if err != nil {
		return false, err
	}
	return true, nil
}

// Count returns how many rows share nodeID (0 or 1 for a healthy log).
func (l *NodeLog) Count(nodeID string) (int, error) {
	l.mu.Lock()
	defer l.mu.Unlock()

	var n int
	err := l.db.QueryRow(`SELECT COUNT(*) FROM nodes WHERE id = ?`, nodeID).Scan(&n)
	return n, err
}

// ListNodes returns nodes for a trajectory scoped to tenantID. tenantID is
// required so this method always enforces tenant isolation at the read path.
// Callers that genuinely need rows across tenants (e.g. tir.Export's own
// mixed-tenant check) must use ListNodesAllTenants explicitly instead.
func (l *NodeLog) ListNodes(trajectoryID, tenantID string) ([]map[string]any, error) {
	return l.listNodes(trajectoryID, &tenantID)
}

// ListNodesAllTenants returns nodes for a trajectory across all tenants.
// This bypasses tenant isolation and should only be used by code that
// performs its own tenant-scoping check on the result, such as tir.Export
// detecting a trajectory with mixed tenant_id values.
func (l *NodeLog) ListNodesAllTenants(trajectoryID string) ([]map[string]any, error) {
	return l.listNodes(trajectoryID, nil)
}

func (l *NodeLog) listNodes(trajectoryID string, tenantID *string) ([]map[string]any, error) {
	l.mu.Lock()
	defer l.mu.Unlock()

	q := `SELECT id, trajectory_id, tenant_id, step_n, seq, kind, payload_json, ts
	      FROM nodes WHERE trajectory_id = ?`
	args := []any{trajectoryID}
	if tenantID != nil {
		q += ` AND tenant_id = ?`
		args = append(args, *tenantID)
	}
	q += ` ORDER BY COALESCE(step_n, -1), seq`

	rows, err := l.db.Query(q, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	out := make([]map[string]any, 0)
	for rows.Next() {
		var (
			id, traj, tenant, kind, payloadJSON string
			stepN                               sql.NullInt64
			seq                                 int
			ts                                  float64
		)
		if err := rows.Scan(&id, &traj, &tenant, &stepN, &seq, &kind, &payloadJSON, &ts); err != nil {
			return nil, err
		}
		var payload map[string]any
		if err := json.Unmarshal([]byte(payloadJSON), &payload); err != nil {
			return nil, err
		}
		rec := map[string]any{
			"id":            id,
			"trajectory_id": traj,
			"tenant_id":     tenant,
			"seq":           seq,
			"kind":          kind,
			"payload":       payload,
			"ts":            ts,
		}
		if stepN.Valid {
			rec["step_n"] = int(stepN.Int64)
		} else {
			rec["step_n"] = nil
		}
		out = append(out, rec)
	}
	return out, rows.Err()
}

// Close releases the database handle.
func (l *NodeLog) Close() error {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.db == nil {
		return nil
	}
	err := l.db.Close()
	l.db = nil
	return err
}
